scheduler/pkg/server/pipeline_status_test.go

```diff
@@ -210,3 +211,276 @@ func TestPipelineStatusEvents(t *testing.T) {
 		})
 	}
 }
+
+func TestPipelineGwRebalanceMessage(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	type test struct {
+		name           string
+		loadReq        *pb.Pipeline
+		pipelineStatus pipeline.PipelineStatus
+	}
+
+	tests := []test{
+		{
+			name: "rebalance pipeline ready",
+			loadReq: &pb.Pipeline{
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*pb.PipelineStep{
+					{
+						Name: "a",
+					},
+					{
+						Name:   "b",
+						Inputs: []string{"a.outputs"},
+					},
+				},
+			},
+			pipelineStatus: pipeline.PipelineReady,
+		},
+		{
+			name: "rebalance pipeline create",
+			loadReq: &pb.Pipeline{
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*pb.PipelineStep{
+					{
+						Name: "a",
+					},
+					{
+						Name:   "b",
+						Inputs: []string{"a.outputs"},
+					},
+				},
+			},
+			pipelineStatus: pipeline.PipelineCreate,
+		},
+		{
+			name: "rebalance pipeline creating",
+			loadReq: &pb.Pipeline{
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*pb.PipelineStep{
+					{
+						Name: "a",
+					},
+					{
+						Name:   "b",
+						Inputs: []string{"a.outputs"},
+					},
+				},
+			},
+			pipelineStatus: pipeline.PipelineCreating,
+		},
+		{
+			name: "rebalance pipeline terminating",
+			loadReq: &pb.Pipeline{
+				Name:    "foo",
+				Version: 1,
+				Uid:     "x",
+				Steps: []*pb.PipelineStep{
+					{
+						Name: "a",
+					},
+					{
+						Name:   "b",
+						Inputs: []string{"a.outputs"},
+					},
+				},
+			},
+			pipelineStatus: pipeline.PipelineTerminating,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// create a test scheduler - note it uses a load balancer with 1 partition
+			s, _ := createTestScheduler()
+
+			// create pipelinegw stream
+			stream := newStubPipelineStatusServer(1, 5*time.Millisecond)
+			subscription := &PipelineSubscription{
+				name:   "dummy",
+				stream: stream,
+				fin:    make(chan bool),
+			}
+			s.pipelineEventStream.streams[stream] = subscription
+			g.Expect(s.pipelineEventStream.streams[stream]).ToNot(BeNil())
+
+			// add stream to the load balancer
+			s.pipelineGWLoadBalancer.AddServer(subscription.name)
+
+			// add a pipeline to the store
+			err := s.pipelineHandler.AddPipeline(test.loadReq)
+			g.Expect(err).To(BeNil())
+
+			err = s.pipelineHandler.SetPipelineState(
+				test.loadReq.Name,
+				test.loadReq.Version,
+				test.loadReq.Uid,
+				test.pipelineStatus,
+				"dummy_reason",
+				"dummy_source",
+			)
+			g.Expect(err).To(BeNil())
+
+			// trigger rebalance and allow events to propagate
+			s.pipelineGwRebalance()
+			time.Sleep(500 * time.Millisecond)
+
+			var msr *pb.PipelineStatusResponse
+			select {
+			case next := <-stream.msgs:
+				msr = next
+			default:
```

Comment:
For new tests that wait for event propagation, I recommend the following pattern:

* sleep for a relatively short duration to allow for event propagation (the one above at 500ms is just fine)
* in the select, rather than having a `default` case (which is selected immediately if no event is present in another case), have a second case with a longer timeout (i.e `case <- time.After(2*second)`. This can prevent flaky tests, especially when running under constrained resources (like CI).